### ISSUE TITLE
Batched gradient computation can spam warnings

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#51144]
## 🐛 Bug

Batched gradient computation warns when a batching rule is not implemented. However, warnings during backward execution may not be turned into Python warnings, leading to console spam (#50209)

## To Reproduce

```py
import torch
from torch._vmap_internals import vmap
import warnings
warnings.simplefilter('ignore')

device = 'cuda:0'
x = torch.randn(4, requires_grad=True, device=device)
indices = torch.tensor([0, 1], device=device)
y = x.gather(0, indices)

def vjp(v):
    return torch.autograd.grad(y, x, v, retain_graph=True)

vs = torch.randn(2, *y.shape, device=device)
vmap(vjp)(vs)
vmap(vjp)(vs)
vmap(vjp)(vs)

# This script has the following output, despite the warnings filter
# [W BatchedFallback.cpp:63] Warning: Batching rule not implemented for aten::gather_backward falling back to slow (for loop and stack) implementation (function batchedTensorForLoopFallback)
# [W BatchedFallback.cpp:63] Warning: Batching rule not implemented for aten::gather_backward falling back to slow (for loop and stack) implementation (function batchedTensorForLoopFallback)
# [W BatchedFallback.cpp:63] Warning: Batching rule not implemented for aten::gather_backward falling back to slow (for loop and stack) implementation (function batchedTensorForLoopFallback)
```

## Challenges

I don't think we are going to fix #50209 anytime soon. Furthermore, using `TORCH_WARN_ONCE` doesn't solve the problem because all operators that go through the fallback use the same exact vmap warning generation code; `TORCH_WARN_ONCE` would cause the warning to appear only at the first operator that raises the warning.

Any solution must take into account that autograd is multi-threaded.

## Potential solutions

Here are a few different proposals for this, categorized into "short-term" and "long-term". I am not suggesting that we do something in the short-term bucket first.

"Short-term"
- Implement a VMAP_WARN_ONCE that allocates a hash map on a heap. If an operator is not in the hash map, then we print a warning and add it to the map. If it is in the map, then we do not print the warning. (Is adding elements to `std::unordered_map` thread safe?)
- By default, disable the vmap fallback warning. Add a private `_ShowVmapWarnings` context manager that enables the vmap fallback warning. We already have [a mechanism to handle propagating thread-local state through autograd threads](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ThreadLocalState.h)

"long term"
- Have an actual "TORCH_WARN_ONCE_PER_PARAMETERS" that isn't some slow implementation around a hashmap. This is probably physically impossible
- (Leveraging a proposal from @ezyang) Have a "performance warnings context manager". When enabled, PyTorch prints on performance warnings. If we had this, then we would classify vmap fallback warnings as "performance warnings".

## Environment

master

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer
